doc/fwutil/fwutil.md

```diff
@@ -0,0 +1,441 @@
+
+In order to improve scalability and performance a modern network switches provide different architecture solutions:
+1. Non modular chassis platforms
+2. Modular chassis platforms
+
+Non modular chassis platforms may contain only one chassis.  
+A chassis may contain it's own set of components.
+
+Modular chassis platforms may contain only one chassis.  
+A chassis may contain one or more modules and it's own set of components.  
+Each module may contain it's own set of components.
+
+Basically each chassis/module may contain one or more components (e.g., BIOS/ASIC/CPLD/FPGA).
+
+SONiC platform API provides an interface for FW maintenance operations for both modular and  
+non modular chassis platforms. Both modular and non modular chassis platforms share the same platform API,  
+but may have different implementation.
+
+SONiC FW utility uses platform API to interact with the various platform components.
+
+## 2.2 FW utility
+
+### 2.2.1 Command structure
+
+**User interface**:
+```
+fwutil
+|--- show
+|    |--- status
+|    |--- version
+|--- install
+|    |--- chassis <chassis_name>
+|         |--- component <component_name>
+|         |    |--- fw -y|--yes <fw_path>
+|         |--- module <module_name>
+|              |--- component <component_name>
+|                   |--- fw -y|--yes <fw_path>
+|--- update -y|--yes -f|--force -i|--image=<current|next>
+```
+
+**Note:** <fw_path> can be absolute path or URL
+
+### 2.2.2 Command interface
+
+#### 2.2.2.1 Show commands
+
+##### 2.2.2.1.1 Overview
+
+The purpose of the show commands group is to provide an interface for:
+1. FW utility related information query (version, etc.)
+2. Platform components related information query (fw, etc.)
+
+##### 2.2.2.1.2 Description
+
+**The following command displays FW utility version:**
+```bash
+root@sonic:~# fwutil show version
+fwutil version 1.0.0.0
+```
+
+**The following command displays platform components and FW versions:**
+1. Non modular chassis platform
+```bash
+root@sonic:~# fwutil show status
+Chassis   Module   Component   Version             Description
+--------  -------  ----------  --------            ------------
+Chassis1  N/A      BIOS        0ACLH003_02.02.007  Chassis BIOS
+                   CPLD        5                   Chassis CPLD
+                   FPGA        5                   Chassis FPGA
+```
+
+2. Modular chassis platform
+```bash
+root@sonic:~# fwutil show status
+Chassis   Module   Component   Version             Description
+--------  -------  ----------  --------            ------------
+Chassis1           BIOS        0ACLH004_02.02.007  Chassis BIOS
+                   CPLD        5                   Chassis CPLD
+                   FPGA        5                   Chassis FPGA
+          Module1  ASIC        29.2000.1886        Module ASIC
+                   CPLD        5                   Module CPLD
+                   FPGA        5                   Module FPGA
+```
```

Comment:
Can the design validate sw compatibility between different components? ex: BIOS & CPLD

Reply:
I don’t think such is correct. I think the update of a component should be as a standalone. If there is a validation per component as discussed should be part of the platform api implementation. Such as crc, etc.